portfolio/src/main/java/com/google/sps/servlets/ImageUploadFormHandlerServlet.java

```diff
@@ -0,0 +1,77 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/image-upload-form-handler")
+public class ImageUploadFormHandlerServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String imageUrl = getUploadedFileUrl(request, "image");
+
+    PrintWriter out = response.getWriter();
+    out.println("<p>Here's the image you uploaded:</p>");
+    out.println("<a href=\"" + imageUrl + "\">");
+    out.println("<img src=\"" + imageUrl + "\" />");
+    out.println("</a>");
+  }
+
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get("image");
+
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
```

Comment:
It's almost never preferable to return null for a function. If anything, you can use [`Optional`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html). 

In this case, it's probably preferable to throw an exception, because this shouldn't happen and if it does we don't want to make things more confusing for a client calling this servlet to debug and would rather give them a nice, short-but-helpful error message they can act upon.

I'd also recommend checking that the key is present using `blobs.containsKey("image")` **before** calling `blobs.get`, because you're assigning it to a variable which isn't marked as `@Nullable` but could be null. By checking with `containsKey` first, you're verifying that `blobKeys` is in fact of type `List<BlobKey>`.

Reply:
So you'd have two different error messages here. One for if there wasn't an image, and another for if the blobKeys is empty.

Reply:
I added some error messages and change the nulls to empty strings. I added a javadoc to explain this function's outputs